src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/CSharpTagHelperCodeRenderer.cs

```diff
@@ -448,7 +448,15 @@ private void RenderTagOutput(string tagOutputMethodName)
                 return;
             }
 
-            CSharpCodeVisitor.RenderPreWriteStart(_writer, _context);
+            // If its a string WriteLiteral must be generated.
```

Comment:
should never generate calls to `Write[To]` here.  previous code was correct because everything `TagHelperOutput` generates is already HTML encoded as necessary.  (`Write` versus `WriteLiteral` has nothing to do with the datatype.  `Write` HTML encodes while `WriteLiteral` does not.)

this should mean many fewer changes to the tests using generated code :smiley:

separately, the MVC changes should also not affect the `Write` versus `WriteLiteral` distinction.  only need to special-case `WriteLiteral[To]` to handle `ITextWriterCopyable`.